src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -78,20 +78,13 @@ private void StopRaisingEvents()
                 {
                     // Stop the RunLoop and wait for the thread to exit gracefully
                     Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
-                    _watcherThread.Join();
-                    _watcherRunLoop = IntPtr.Zero;
                 }
 
-                // Clean up the EventStream, if it exists
-                if (!_eventStream.IsInvalid)
-                {
-                    _eventStream = new SafeEventStreamHandle(IntPtr.Zero);
-                }             
-
-                // Cleanup the callback
-                if (_callback != null)
+                // Gurd against us from calling Join on our own background thread (for 
+                // example, calling Dispose or Stop from a callback).
+                if (Thread.CurrentThread != _watcherThread)
                 {
-                    _callback = null;
+                    _watcherThread.Join();
```

Comment:
My concern with the Join is that it's introducing a dependency between threads that's not there in the other platform implementations.  That means code which doesn't deadlock on one platform is prone to on another.  Even though it's unlikely, a deadlock could happen if the callback tries to take a lock that's held by another thread that's tried to stop raising events.

Coudl we do something like what we do in the Linux implementation, where basically all of the state is kept in a separate object, and the only connection between the main object and the state object is a cancellation token source that's used to shut things down?  Or something like that.

If such a change will take more time, I'm ok if we want to merge this change as a temporary workaround for the deadlock that's causing even bigger problems, but we should definitely address the corruption potential.